be/src/exec/pipeline/scan_operator.cpp

```diff
@@ -8,177 +8,134 @@
 }
 
 bool ScanOperator::is_finished() const {
-    return _is_finished;
-}
-
-void ScanOperator::finish(RuntimeState* state) {
-    _is_finished = true;
-    if (_chunk_source) {
-        _chunk_source->close(state);
+    if (_is_finished) {
+        return true;
     }
-}
 
-StatusOr<vectorized::ChunkPtr> ScanOperator::_pull_chunk_blocking(RuntimeState* state) {
-    DCHECK(_io_threads == nullptr);
-    if (_is_finished) {
-        return Status::EndOfFile("End-Of-Stream");
+    if (_has_next_morsel) {
+        return false;
     }
-    DCHECK(_chunk_source);
-    auto chunk = _chunk_source->get_next_chunk();
-    if (chunk.ok() || !chunk.status().is_end_of_file()) {
-        return chunk;
+
+    // If there is no next morsel, and io task is active
+    // we just wait for the io thread to end during which current pipeline
+    // will still be scheduled multiply times, but it can be guarantee that
+    // the waiting time is very short
+    if (_io_task_state.load(std::memory_order_acquire) == IOTaskState::INACTIVE) {
+        _is_finished = true;
     }
-    _pickup_morsel(state);
-    return std::make_shared<vectorized::Chunk>();
+
+    return _is_finished;
 }
 
-StatusOr<vectorized::ChunkPtr> ScanOperator::_pull_chunk_nonblocking(RuntimeState* state) {
-    DCHECK(_io_threads != nullptr);
-    if (_is_finished) {
-        return Status::EndOfFile("End-Of-Stream");
+StatusOr<vectorized::ChunkPtr> ScanOperator::pull_chunk(RuntimeState* state) {
+    DCHECK(_chunk_source != nullptr);
+    if (!_chunk_source->has_output()) {
+        if (_chunk_source->has_next_chunk()) {
+            _trigger_next_scan();
+        } else {
+            _pickup_morsel(state);
+        }
+        return nullptr;
     }
 
-    _chunk_source = _pending_chunk_source_future.value().get();
-    _pending_chunk_source_future = {};
-    auto chunk = _chunk_source->get_next_chunk_nonblocking();
-    if (chunk.ok()) {
-        _trigger_read_chunk();
-        return chunk;
-    }
-    if (!chunk.status().is_end_of_file()) {
-        return chunk;
-    }
-    // Now ScanOperator can process multiple morsels, when the non-last morsel is
-    // processed and the EndOfFile is encountered, then ScanOperator has no chunk
-    // to output and should pick up next morsel. so here return nullptr instead of
-    // empty chunk.
-    _pickup_morsel(state);
-    return std::make_shared<vectorized::Chunk>();
+    return _chunk_source->get_next_chunk_from_cache();
 }
 
-StatusOr<vectorized::ChunkPtr> ScanOperator::pull_chunk(RuntimeState* state) {
-    if (_io_threads == nullptr) {
-        return _pull_chunk_blocking(state);
+void ScanOperator::_trigger_next_scan() {
+    DCHECK(_io_task_state.load(std::memory_order_acquire) == IOTaskState::INACTIVE);
+
+    PriorityThreadPool::Task task;
+    _io_task_state.store(IOTaskState::PENDING, std::memory_order_release);
+    task.work_function = [this]() {
+        _io_task_state.store(IOTaskState::ACTIVE, std::memory_order_release);
```

Comment:
rename std::memory_order_release to default memory order. since memory_order_release is used to prevent preceding independent write operations is moved after memory_order_release. so the following independent write/read is allowed to moved before.

Reply:
fixed